.circleci/integration-tests/airflow_dag_introspection.py

```diff
@@ -0,0 +1,174 @@
+                # Wait for some time before retrying
+                time.sleep(60)
+                print("Retrying...")
+                # Increment the number of retries
+                retries += 1
+            else:
+                print("Maximum number of retries reached.")
+    return logs
+
+
+def log_checker(ti_id: str, expected: str, notexpected: str, try_number: int = 1, **context: dict):
+    """Get Task logs of a given TI, also validate expected substring,
+    we can also provide a substring which is not expected
+    """
+    time.sleep(30)
+    dagrun = context["dag_run"]
+    task_instances = dagrun.get_task_instances()
+    this_task_instance = next(filter(lambda ti: ti.task_id == ti_id, task_instances))  # ti_call
+    context["dag"].dag_id
+    context["dag_run"].run_id
+
+    def check(ti, expect, notexpect):
+        with create_session() as session:
+            ti = (
+                session.query(TaskInstance)
+                .filter(
+                    TaskInstance.task_id == ti.task_id,
+                    TaskInstance.dag_id == ti.dag_id,
+                    TaskInstance.run_id == ti.run_id,
+                    TaskInstance.map_index == ti.map_index,
+                )
+                .join(TaskInstance.dag_run)
+                .options(joinedload("trigger"))
+                .options(joinedload("trigger.triggerer_job"))
+            ).first()
+        task_log_reader = TaskLogReader()
+
+        log_container, _ = task_log_reader.read_log_chunks(ti, try_number, {"download_logs": True})
+        logs = log_checker_with_retry(10, log_container)
+        print(f"Found logs: {logs}")
+        assert notexpect not in logs  # nosec: B101
+        assert expect in logs  # nosec: B101
+        print(f"Found {expect} but not {notexpect}")
+
+    # make sure expected output appeared
+    check(this_task_instance, expected, notexpected)
+
+
+def assert_homomorphic(task_group_names, **context):
+    """The structure of all of the task groups above should be the same"""
+    # get the dag in dot notation, focus only on its edges
+    dag = context["dag"]
+    print(dag)
+    # gives string which represents whole dag structure
+    graph = render_dag(dag)
+    print("Whole DAG:")
+    print(indent(str(graph), "    "))
+    lines = list(filter(lambda x: "->" in x, str(graph).split("\n")))
+
+    # bin them by task group, then remove the group names
+    group_strings = []
+    # removes everything that's not a task name
+    for name in task_group_names:
+        print(name)
+        relevant_lines = filter(lambda x: name in x, lines)
+        normalized_lines = (x.strip().replace(name, "") for x in sorted(relevant_lines))
+        edges_str = "\n".join(normalized_lines)
+        group_strings.append(edges_str)
+        print(indent(edges_str, "    "))
+
+    # these should be identical
+    for xgroup, ygroup in zip(group_strings, group_strings[1:]):
+        assert xgroup == ygroup  # nosec: B101
```

Comment:
After a second thought, as we're actually running these lines on astro cloud, should we raise another exception when these conditions aren't met instead of using assert?